### PR TITLE
[cxx-interop] Fix spurious ambiguous member lookup

### PR DESCRIPTION
**Pull request by j-hui**
Nested calls to `importBaseMemberDecl()` subvert its cache and compromise its idempotence, causing the semantic checker to spuriously report ambiguous member lookups when multiple `ClangRecordMemberLookup` requests are made (e.g., because of an unrelated missing member lookup).

One such scenario is documented in the test case I checked into the test suite in f818a5d72a595c90bf2f3672960081ed495f6f7e, which fails without this fix (in particular, `inherited-lookup-typechecker.swift` fails because of the expected error from the `missing` member; `inherited-lookup-executable.swift` works because it does not attempt to access a `missing` member).

I noticed this bug in #77726 when I found that my `expected-error`s seemed to be triggering this unrelated error. I'm breaking out a small piece of that PR because (1) it is getting quite large, (2) it is getting stale, and (3) I am trying to separate this bug fix from that feature work.

rdar://141069984

### CODE REVIEW

**Comment by j-hui** [comment]
Another reason I'm opening this PR separate from #77726 is because I want to test it independently in CI. This change seems to cause a cascade of other issues, apparently because the existing tests have come to depend on nested calls to `importBaseMemberDecl()`; each layer of nesting synthesizes another (unnecessary) shim call, with an even longer identifier.

As such, @swift-ci please test

**Comment by j-hui** [comment]
@swift-ci please test

**Comment by j-hui** [comment]
@swift-ci Please smoke test

**Comment by j-hui** [comment]
@swift-ci Please test

**Comment by j-hui** [comment]
@swift-ci Please test

File: test/Interop/Cxx/class/inheritance/Inputs/inherited-lookup.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+struct Base1 {
+  int method(void) const { return 1; }
+};
+
+struct IBase1 : Base1 {};
```
Comment (Xazax-hun) [inline test/Interop/Cxx/class/inheritance/Inputs/inherited-lookup.h:7]: Could we add a method at each layer and test that all of them are properly imported?

File: lib/ClangImporter/ClangImporter.cpp [hunk 2]
```diff
@@ -6197,9 +6199,11 @@ TinyPtrVector<ValueDecl *> ClangRecordMemberLookup::evaluate(
           continue;
 
         // Add Clang members that are imported lazily.
-        auto baseResults = evaluateOrDefault(
-            ctx.evaluator,
-            ClangRecordMemberLookup({cast<NominalTypeDecl>(import), name}), {});
+        auto baseResults =
```
Comment (Xazax-hun) [inline lib/ClangImporter/ClangImporter.cpp:6202]: So `baseResults` includes all the members from all the bases recursively here?

File: lib/ClangImporter/ClangImporter.cpp [hunk 3]
```diff
@@ -6218,6 +6222,21 @@ TinyPtrVector<ValueDecl *> ClangRecordMemberLookup::evaluate(
           // as that would cause an ambiguous lookup.
           if (foundNameArities.count(getArity(foundInBase)))
             continue;
+
+          // Do not importBaseMemberDecl() if this is a recursive lookup into
+          // some class's superclass. importBaseMemberDecl() caches synthesized
+          // members, which does not work if we call it on its result, e.g.:
+          //
+          //    importBaseMemberDecl(importBaseMemberDecl(foundInBase,
```
Comment (Xazax-hun) [inline lib/ClangImporter/ClangImporter.cpp:6230]: Could we still run into a similar problem if the same `recordDecl` is accessible on multiple paths in the inheritance DAG? E.g.:
```
struct A {};
struct B : A {};
struct C : B, A {};
```

File: lib/ClangImporter/ClangImporter.cpp (hunk 2 above)
Comment (j-hui) [reply lib/ClangImporter/ClangImporter.cpp:6202]: It includes all the results inherited from only `import` here. This is inside of a loop that iterates through all base classes, so `baseResults` will contain the results for each base class (and its ancestors) separately.

**Comment by j-hui** [comment]
@swift-ci Please test

File: test/Interop/Cxx/class/inheritance/Inputs/inherited-lookup.h (hunk 1 above)
Comment (j-hui) [reply test/Interop/Cxx/class/inheritance/Inputs/inherited-lookup.h]: I've added a method at each layer, although I do believe there is already quite a lot of coverage for that in the rest of the Swift test suite.

My intention here is to ensure coverage for a very specific scenario I encountered while studying the implementation of `ClangRecordMemberLookup`, at least for now. Later, I plan on expanding this test suite in subsequent patches to handle issues like overridden methods and ambiguous lookups (which you had foreshadowed in another comment). These patches will all be necessary to build back up to https://github.com/swiftlang/swift/pull/77726 in an incremental and robust way.

**Review by Xazax-hun — APPROVED** [review]
LGTM!

**Review by fahadnayyar — APPROVED** [review]

**Pull request merged**